### PR TITLE
[16.0][IMP][FIX] stock_move_location: improve planned transfer

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
Before this commit, when a planned transfer for a product with a specific lot number was made, it did not adhere to this lot number and assigned a different lot as per Odoo's standard behavior.

After this commit, planned transfers will respect the lot number and create move lines with the chosen lots.

@qrtl

### CODE REVIEW

**Comment by rafaelbn** [comment]
Please @AungKoKoLin1997 could you rebase, the new readme is merged 😄

**Review by rinaldifirdaus — APPROVED** [review]
Functional test: Work as Expected
for sample, adjust 3 units with lot number each and store it to WH/Stock.
Added Lot:
- CM-1 (Firstly added lot)
- CM-2
- CM-3

Move CM-2 and CM-3 to other location using move location wizard and click Planned transfer.
And the transfer that created reserved the lot CM-2 and CM-3 properly. 

image
image
image

**Review by rinaldifirdaus — APPROVED** [review]

File: stock_move_location/wizard/stock_move_location_line.py [hunk 1]
```diff
@@ -92,6 +92,21 @@ def create_move_lines(self, picking, move):
...
+                    strict=False,
+                )
+                # Create stock.move.line
+                move._update_reserved_quantity(
+                    values.get("qty_done"),
+                    available_quantity,
+                    line.origin_location_id,
+                    lot_id=line.lot_id,
+                    strict=False,
+                )
```
Comment (yostashiro) [inline stock_move_location/wizard/stock_move_location_line.py]: It's a bit hard to understand the context of this additoon. Can you please try and find a way to write on created move lines, as that should keep records consistent (and we should be able to remove this part).

File: stock_move_location/wizard/stock_move_location.py [hunk 2]
```diff
@@ -207,9 +209,13 @@ def _get_move_values(self, picking, lines):
     def _create_move(self, picking, lines):
         self.ensure_one()
         move = self.env["stock.move"].create(self._get_move_values(picking, lines))
-        if not self.env.context.get("planned"):
-            for line in lines:
-                line.create_move_lines(picking, move)
+        for line in lines:
+            line.create_move_lines(picking, move)
```
Comment (yostashiro) [inline stock_move_location/wizard/stock_move_location.py]: `create_move_lines()` works on a recordset.

```suggestion
        lines.create_move_lines(picking, move)
```

File: stock_move_location/wizard/stock_move_location.py [hunk 3]
```diff
@@ -207,9 +209,13 @@ def _get_move_values(self, picking, lines):
...
+        if self.env.context.get("planned"):
```
Comment (yostashiro) [inline stock_move_location/wizard/stock_move_location.py:213]: To have the logic consolidated in one place for 'planned' transfers.

```suggestion
        if self.env.context.get("planned"):
            for line in lines:
                available_quantity = self.env["stock.quant"]._get_available_quantity(
                    line.product_id,
                    line.origin_location_id,
                    lot_id=line.lot_id,
                    strict=False,
                )
                move._update_reserved_quantity(
                    line.move_quantity,
                    available_quantity,
                    line.origin_location_id,
                    lot_id=line.lot_id,
                    strict=False,
                )
```

File: stock_move_location/wizard/stock_move_location_line.py [hunk 4]
```diff
@@ -92,6 +92,21 @@ def create_move_lines(self, picking, move):
...
+                # Update move line accordingly.
+                move._update_reserved_quantity(
+                    values.get("qty_done"),
+                    available_quantity,
+                    line.origin_location_id,
+                    lot_id=line.lot_id,
+                    strict=False,
+                )
```
Comment (yostashiro) [inline stock_move_location/wizard/stock_move_location_line.py]:
```suggestion
```

File: stock_move_location/wizard/stock_move_location_line.py [hunk 5]
```diff
@@ -124,9 +139,6 @@ def _get_available_quantity(self):
         self.ensure_one()
         if not self.product_id:
             return 0
-        if self.env.context.get("planned"):
-            # for planned transfer we don't care about the amounts at all
-            return self.move_quantity, 0
```
Comment (yostashiro) [inline stock_move_location/wizard/stock_move_location_line.py]: @AungKoKoLin1997 As we just discussed, please revert this.

File: stock_move_location/wizard/stock_move_location.py [hunk 6]
```diff
@@ -256,8 +276,7 @@ def action_move_location(self):
             picking.button_validate()
             moves_to_reassign._action_assign()
         else:
-            picking.action_confirm()
-            picking.action_assign()
+            picking.state = "assigned"
```
Comment (yostashiro) [inline stock_move_location/wizard/stock_move_location.py]: This doesn't look right. I think we should just remove the else clause.

File: stock_move_location/wizard/stock_move_location.py [hunk 7]
```diff
@@ -207,9 +209,27 @@ def _get_move_values(self, picking, lines):
...
+                    strict=False,
+                )
+                move._update_reserved_quantity(
+                    line.move_quantity,
+                    available_quantity,
+                    line.origin_location_id,
+                    lot_id=line.lot_id,
+                    strict=False,
+                )
+
```
Comment (yostashiro) [inline stock_move_location/wizard/stock_move_location.py]:
```suggestion
```

File: stock_move_location/readme/USAGE.md [hunk 8]
```diff
@@ -12,6 +12,8 @@
   show up
 - If "PLANNED TRANSFER" is used - the picking won't be validated
   automatically
+- When you choose specific lot/serial numbers to be moved in the wizard and opt for 'PLANNED TRANSFER,' 
+  the selected lot/serial will be reserved on the Picking.
```
Comment (yostashiro) [inline stock_move_location/readme/USAGE.md]: Let's remove this comment; it's a bit incomplete and random.
```suggestion
```

**Review by yostashiro — APPROVED** [review]
LGTM.

**Comment by yostashiro** [comment]
Just added another fix in a separate commit: fixed `_get_locations_domain()` to use the correct company under multi-company settings.

**Comment by AungKoKoLin1997** [comment]
@OCA-stock-logistics-warehouse-maintainers 
Is it possible to look at this PR?

File: stock_move_location/wizard/stock_move_location.py [hunk 9]
```diff
@@ -207,9 +209,26 @@ def _get_move_values(self, picking, lines):
...
+                    line.product_id,
+                    line.origin_location_id,
+                    lot_id=line.lot_id,
+                    strict=False,
+                )
+                move._update_reserved_quantity(
+                    line.move_quantity,
+                    available_quantity,
+                    line.origin_location_id,
+                    lot_id=line.lot_id,
```
Comment (rousseldenis) [inline stock_move_location/wizard/stock_move_location.py:227]: @AungKoKoLin1997 You should also take into account owner and package fields in both methods

**Review by rousseldenis — CHANGES_REQUESTED** [review]
LGTM. Apart remarks on missing arguments

File: stock_move_location/wizard/stock_move_location.py (hunk 9 above)
Comment (AungKoKoLin1997) [reply stock_move_location/wizard/stock_move_location.py:227]: Thanks for the review.  I updated the code.

**Review by rousseldenis — APPROVED** [review]

**Comment by AungKoKoLin1997** [comment]
@rousseldenis Is it possible to get this PR merged?

**Pull request merged**